spark/src/main/scala/org/apache/spark/sql/delta/actions/TableFeatureSupport.scala

```diff
@@ -258,24 +270,32 @@ trait TableFeatureSupport { this: Protocol =>
   def canDowngradeTo(to: Protocol, droppedFeatureName: String): Boolean = {
     if (!supportsWriterFeatures) return false
 
-    // When `to` protocol does not have any features version downgrades are possible. However,
+    val featureNames = readerAndWriterFeatureNames - droppedFeatureName
+    val minRequiredVersions = TableFeatureProtocolUtils.minimumRequiredVersions(
+      featureNames.flatMap(TableFeature.featureNameToFeature).toSeq)
+    val toSupportsRequiredMinVersions =
+      (to.minReaderVersion, to.minWriterVersion) == minRequiredVersions
+
+    // When `to` protocol does not have any features, versions might get downgraded. However,
     // the current protocol needs to contain one non-legacy feature. We also allow downgrade when
     // there are only legacy features. This is to accommodate the case when the user attempts to
     // remove a legacy feature in a table that only contains legacy features.
     if (to.readerAndWriterFeatureNames.isEmpty) {
-      val featureNames = readerAndWriterFeatureNames - droppedFeatureName
       val sameLegacyFeaturesSupported = featureNames == to.implicitlySupportedFeatures.map(_.name)
-      val minRequiredVersions = TableFeatureProtocolUtils.minimumRequiredVersions(
-        featureNames.flatMap(TableFeature.featureNameToFeature).toSeq)
 
       return sameLegacyFeaturesSupported &&
         (to.minReaderVersion, to.minWriterVersion) == minRequiredVersions &&
         readerAndWriterFeatures.filterNot(_.isLegacyFeature).size <= 1
     }
 
-    // When `to` protocol contains table features we cannot downgrade the protocol version.
-    if (to.minReaderVersion != this.minReaderVersion) return false
-    if (to.minWriterVersion != this.minWriterVersion) return false
+    // When `to` protocol contains table features, protocol versions may still
+    // get downgraded. However, the resulting protocol needs to support at least writer features.
+    // Note, we check the minimum required versions only when table features exist in the protocol.
+    // This because we do not always downgrade the versions of protocols with no table features.
+    if (!to.supportsWriterFeatures ||
```

Comment:
Please use parenteses when combining `||` and `&&` -- I always forget the precedence order, and it's easy to get it wrong.